### PR TITLE
Bump upper bound on containers to <0.7

### PR DESCRIPTION
**Pull request by bgamari**
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Just as it says on the tin. Note that I had to do a bit of import-wrangling to get this to compile as `Data.Map.insertWith'` is now gone.

### CODE REVIEW

**Comment by bgamari** [comment]
It doesn't look like the one failure is due to this patch.

File: Cabal/Distribution/PackageDescription/Configuration.hs [hunk 1]
```diff
@@ -63,7 +63,8 @@ import Distribution.Types.CondTree
 import Distribution.Types.Condition
 import Distribution.Types.DependencyMap
 
-import qualified Data.Map as Map
+import qualified Data.Map.Strict as Map.Strict
```
Comment (phadej) [inline Cabal/Distribution/PackageDescription/Configuration.hs]: this module doesn't exist in `containers-0.4`. We have `Cabal.Distribution.Compat.Map` to deal with that.

Comment (bgamari) [reply Cabal/Distribution/PackageDescription/Configuration.hs]: Actually, it looks like `Data.Map.Lazy` also doesn't exist yet there is not a compatibility wrapper for this. I suppose one should be introduced.

Comment (phadej) [reply Cabal/Distribution/PackageDescription/Configuration.hs]: Yes. I think there we need the lazy variant too.

**Comment by phadej** [comment]
Note: we'll need to update bootstrap script too.

**Comment by bgamari** [comment]
> Note: we'll need to update bootstrap script too.

I checked the bootstrap script and saw no particular reference to `containers`. I suspect it takes advantage of the fact that `containers` is GHC core library.

**Pull request merged**